### PR TITLE
Base 820 refactoring

### PR DESCRIPTION
**Pull request by antbat**

### CODE REVIEW

File: src/main/kotlin/com/bitclave/node/configuration/gson/PageWithCountersSerializer.kt [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+import java.lang.reflect.Type
+
+class PageWithCountersSerializer : JsonSerializer<EnrichedOffersWithCountersResponse> {
+
+    override fun serialize(
+        src: EnrichedOffersWithCountersResponse,
+        typeOfSrc: Type,
+        context: JsonSerializationContext
+    ): JsonElement {
+        val json = PageSerializer().serialize(src, typeOfSrc, context).asJsonObject
```
Comment (AndreyShashlov) [inline src/main/kotlin/com/bitclave/node/configuration/gson/PageWithCountersSerializer.kt]: use context.

File: src/main/kotlin/com/bitclave/node/configuration/gson/PageWithCountersSerializer.kt [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        val counters = JsonObject()
+        src.counters.forEach { (key, value) ->
```
Comment (AndreyShashlov) [inline src/main/kotlin/com/bitclave/node/configuration/gson/PageWithCountersSerializer.kt]: use context.

File: src/main/kotlin/com/bitclave/node/repository/models/controllers/EnrichedOffersWithCountersResponse.kt [hunk 3]
```diff
@@ -2,37 +2,9 @@ package com.bitclave.node.repository.models.controllers
...
-        this.numberOfElements = offers.numberOfElements
-        this.first = offers.isFirst
-        this.last = offers.isLast
-        this.number = offers.number
-        this.size = offers.size
-
-        this.pageable = PageRequest.of(this.number, this.size)
-    }
-}
+class EnrichedOffersWithCountersResponse(
```
Comment (AndreyShashlov) [inline src/main/kotlin/com/bitclave/node/repository/models/controllers/EnrichedOffersWithCountersResponse.kt]: data class + val for arg

File: src/main/kotlin/com/bitclave/node/repository/models/controllers/OffersWithCountersResponse.kt [hunk 4]
```diff
@@ -1,26 +1,18 @@
...
-
-    val total: Long = 0
-    val numberOfElements = 0
-    val first = false
-    val last = false
-    val number = 0
-    val size = 10
-    val totalPages = 0
-    val totalElements: Long = 0
+class OffersWithCountersResponse(
```
Comment (AndreyShashlov) [inline src/main/kotlin/com/bitclave/node/repository/models/controllers/OffersWithCountersResponse.kt]: data class + val for arg

File: src/main/kotlin/com/bitclave/node/configuration/gson/PageWithCountersSerializer.kt (hunk 1 above)
Comment (antbat) [reply src/main/kotlin/com/bitclave/node/configuration/gson/PageWithCountersSerializer.kt]: used context everywhere, where it works
`

    private val tokenForCounters = object : TypeToken<Map<String, Map<String, Int>>>() {}.type

    override fun serialize(
        src: EnrichedOffersWithCountersResponse,
        typeOfSrc: Type,
        context: JsonSerializationContext
    ): JsonElement {
        val json = PageSerializer().serialize(src, typeOfSrc, context).asJsonObject
        json.add("counters",  context.serialize(src.counters, tokenForCounters))
        return json
    }

`

File: src/main/kotlin/com/bitclave/node/configuration/gson/PageWithCountersSerializer.kt (hunk 2 above)
Comment (antbat) [reply src/main/kotlin/com/bitclave/node/configuration/gson/PageWithCountersSerializer.kt]: see above

File: src/main/kotlin/com/bitclave/node/repository/models/controllers/EnrichedOffersWithCountersResponse.kt (hunk 3 above)
Comment (antbat) [reply src/main/kotlin/com/bitclave/node/repository/models/controllers/EnrichedOffersWithCountersResponse.kt]: done.

File: src/main/kotlin/com/bitclave/node/repository/models/controllers/OffersWithCountersResponse.kt (hunk 4 above)
Comment (antbat) [reply src/main/kotlin/com/bitclave/node/repository/models/controllers/OffersWithCountersResponse.kt]: done

File: src/main/kotlin/com/bitclave/node/repository/models/controllers/OffersWithCountersResponse.kt [hunk 5]
```diff
@@ -1,26 +1,18 @@
...
-    val numberOfElements = 0
-    val first = false
-    val last = false
-    val number = 0
-    val size = 10
-    val totalPages = 0
-    val totalElements: Long = 0
+data class OffersWithCountersResponse(
+    val counters: Map<String, Map<String, Int>>,
+    val offerIds: Page<Long>
```
Comment (AndreyShashlov) [inline src/main/kotlin/com/bitclave/node/repository/models/controllers/OffersWithCountersResponse.kt]: make private field offerIds (now we have two same fields: offerIds and content)

File: src/main/kotlin/com/bitclave/node/repository/models/controllers/EnrichedOffersWithCountersResponse.kt [hunk 6]
```diff
@@ -2,37 +2,9 @@ package com.bitclave.node.repository.models.controllers
...
-        this.first = offers.isFirst
-        this.last = offers.isLast
-        this.number = offers.number
-        this.size = offers.size
-
-        this.pageable = PageRequest.of(this.number, this.size)
-    }
-}
+data class EnrichedOffersWithCountersResponse(
+    val offers: Page<OfferSearchResultItem>,
```
Comment (AndreyShashlov) [inline src/main/kotlin/com/bitclave/node/repository/models/controllers/EnrichedOffersWithCountersResponse.kt]: make private field offers (now we have two same fields: offers and content)

File: src/main/kotlin/com/bitclave/node/configuration/gson/PageWithCountersSerializer.kt [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    private val pageTokenType = object : TypeToken<Page<Long>>() {}.type
+    private val tokenForCounters = object : TypeToken<Map<String, Map<String, Int>>>() {}.type
+
+    override fun serialize(
+        src: EnrichedOffersWithCountersResponse,
+        typeOfSrc: Type,
+        context: JsonSerializationContext
+    ): JsonElement {
+        val json = context.serialize(src, pageTokenType).asJsonObject
+        json.add("counters", context.serialize(src.counters, tokenForCounters))
```
Comment (AndreyShashlov) [inline src/main/kotlin/com/bitclave/node/configuration/gson/PageWithCountersSerializer.kt]: separate code for easier reading, like this:
val counters = context.serialize(src.counters, tokenForCounters)
json.add("counters", counters)

File: src/main/kotlin/com/bitclave/node/configuration/gson/PageWithCountersSerializer.kt [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+        return json
```
Comment (AndreyShashlov) [inline src/main/kotlin/com/bitclave/node/configuration/gson/PageWithCountersSerializer.kt:25]: empty single line before return (code style of project)

**Pull request merged**